### PR TITLE
docs: clarify behavior of hedge_on_per_try_timeout

### PR DESCRIPTION
**Pull request by ikonst**

### CODE REVIEW

**Comment by ikonst** [comment]
@snowp I had a hard time understanding the original phrasing so I took a stab at rewriting it, but I'm not really sure I'm describing how it actually works.

**Comment by snowp** [comment]
Thanks for improving the docs!

@mpuncel You wanna give this a look before I do?

@ikonst Can you fix DCO? There are steps to do so in CONTRIBUTING.md

**Comment by ikonst** [comment]
bump @mpuncel

File: api/envoy/api/v2/route/route_components.proto [hunk 1]
```diff
@@ -1128,10 +1128,12 @@ message HedgePolicy {
   // Indicates that a hedged request should be sent when the per-try timeout
   // is hit. This will only occur if the retry policy also indicates that a
   // timed out request should be retried.
-  // Once a timed out request is retried due to per try timeout, the router
-  // filter will ensure that it is not retried again even if the returned
-  // response headers would otherwise be retried according the specified
-  // :ref:`RetryPolicy <envoy_api_msg_route.RetryPolicy>`.
+  // If a response to the hedged-against request arrives, it will be evaluated
+  // according to the :ref:`RetryPolicy <envoy_api_msg_route.RetryPolicy>`.
+  // If per policy it would normally be retried (e.g. status code 409 with
```
Comment (snowp) [inline api/envoy/api/v2/route/route_components.proto:1133]: maybe `per the retry policy` or just `per the policy`?

File: api/envoy/api/v2/route/route_components.proto [hunk 2]
```diff
@@ -1128,10 +1128,12 @@ message HedgePolicy {
...
+  // retriable-4xx policy), it would be discarded since a retry
+  // is already in progress. Otherwise, it would be returned to the caller,
+  // and the hedged request would be aborted and its result discarded.
```
Comment (snowp) [inline api/envoy/api/v2/route/route_components.proto:1136]: The phrasing here kinda implies that you have one request and a hedged version of that request, but in reality you can have many hedges out at the same time. Maybe rephrase it to be more clearer around that? e.g. "all outstanding/hedged requests will be aborted"

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks for improving the docs!

I think you'll also want to update the V3 docs, V2 is on its way out. The V4alpha docs should then be automatically updated once you run the proto_format.sh script.

File: api/envoy/api/v2/route/route_components.proto [hunk 3]
```diff
@@ -1125,13 +1125,13 @@ message HedgePolicy {
...
 
-  // Indicates that a hedged request should be sent when the per-try timeout
-  // is hit. This will only occur if the retry policy also indicates that a
-  // timed out request should be retried.
-  // Once a timed out request is retried due to per try timeout, the router
-  // filter will ensure that it is not retried again even if the returned
-  // response headers would otherwise be retried according the specified
-  // :ref:`RetryPolicy <envoy_api_msg_route.RetryPolicy>`.
+  // Indicates that a hedged request should be sent when the per-try timeout is hit.
+  // This will only occur if the :ref:`RetryPolicy <envoy_api_msg_route.RetryPolicy>` also indicates that
```
Comment (mpuncel) [inline api/envoy/api/v2/route/route_components.proto:1129]: I don't think this part is right, retry policy doesn't matter when per try timeout is hit.

Comment (mpuncel) [reply api/envoy/api/v2/route/route_components.proto:1129]: reference comment: 
```
  // A hedged retry on per try timeout is always retried if there are retries
  // left. NOTE: this is a bit different than non-hedged per try timeouts which
  // are only retried if the applicable retry policy specifies either
  // RETRY_ON_5XX or RETRY_ON_GATEWAY_ERROR. This is because these types of
  // retries are associated with a stream reset which is analogous to a gateway
  // error. When hedging on per try timeout is enabled, however, there is no
  // stream reset.

```

File: api/envoy/api/v2/route/route_components.proto [hunk 4]
```diff
@@ -1125,13 +1125,13 @@ message HedgePolicy {
...
+  // timed out requests should be retried (e.g. retry_on set to 'gateway-error' etc). (Other retry policies
+  // would also apply, but would only have effect if the response came back before the request was hedged against;
+  // otherwise such responses would simply be discarded as a retry is already in flight.)
```
Comment (mpuncel) [inline api/envoy/api/v2/route/route_components.proto:1132]: I think this might be clearer stated as

"Any response received after the timeout and subsequent hedge attempt will never be retried, no matter the RetryPolicy"

File: api/envoy/api/v2/route/route_components.proto [hunk 5]
```diff
@@ -1125,13 +1125,13 @@ message HedgePolicy {
...
+  // With hedging, the first request to complete successfully (be it the original or one of the hedged requests)
+  // wil be the one returned to the caller, where "successfully" is defined as not triggering the retry policy.
```
Comment (mpuncel) [inline api/envoy/api/v2/route/route_components.proto:1134]: typo: will

File: api/envoy/api/v2/route/route_components.proto (hunk 3 above)
Comment (ikonst) [reply api/envoy/api/v2/route/route_components.proto:1129]: Interesting. I was actually working off my experiencing using this feature: I only started seeing (hedged) retries once I set `x-envoy-retry-on`.

File: api/envoy/api/v2/route/route_components.proto (hunk 4 above)
Comment (ikonst) [reply api/envoy/api/v2/route/route_components.proto:1132]: Assuming timeout 150ms, per-try timeout of 50ms, 3 retries and retry-on: 5xx policy, and hedging enabled:

0ms: Request 1 sent.
50ms: Request 1 times out, (hedged) request 2 sent.
75ms: Request 2 (hedged) returns 500.
150ms: Request 1 times out.

Would there be a 3rd request?

File: api/envoy/api/v2/route/route_components.proto (hunk 3 above)
Comment (ikonst) [reply api/envoy/api/v2/route/route_components.proto:1129]: Looking at this code, it seems that `retries_remaining_` would only get initialized to a non-zero value if `retry_on_` is set:
https://github.com/envoyproxy/envoy/blob/2709b6bfc73d089d823c5cde6c883806ec8f3d8e/source/common/router/retry_state_impl.cc#L121-L127

Thus `shouldRetry` would end up returning `RetryStatus::NoRetryLimitExceeded` for the hedged requests.

File: api/envoy/api/v2/route/route_components.proto (hunk 2 above)
Comment (ikonst) [reply api/envoy/api/v2/route/route_components.proto]: reworded

File: api/envoy/api/v2/route/route_components.proto (hunk 1 above)
Comment (ikonst) [reply api/envoy/api/v2/route/route_components.proto]: reworded

File: api/envoy/api/v2/route/route_components.proto (hunk 4 above)
Comment (mpuncel) [reply api/envoy/api/v2/route/route_components.proto:1132]: there would be a 3rd request. Request 2 is considered a new attempt, so it will be retried if it times out or returns a 500. If request 1 comes back with a 500 after request 2 has already been sent, that will be dropped and not retried

Comment (ikonst) [reply api/envoy/api/v2/route/route_components.proto:1132]: ahh, I read `(Any response received after the timeout) AND (subsequent hedge attempt) -> will never be retried` 🤦

File: api/envoy/api/v2/route/route_components.proto [hunk 6]
```diff
@@ -1125,13 +1125,18 @@ message HedgePolicy {
...
-  // response headers would otherwise be retried according the specified
-  // :ref:`RetryPolicy <envoy_api_msg_route.RetryPolicy>`.
+  // Indicates that a hedged request should be sent when the per-try timeout is hit.
+  // This means that a retry will be issued without resetting the original request, leaving multiple upstream requests in flight.
+  // The first request to complete successfully will be the one returned to the caller.
+  //
+  // * At any time, a successful response (i.e. not triggering any of the retry-on conditions) would be returned to the client.
+  // * Before per-try timeout, an error response (per retry-on conditions) would be retried immediately or returned ot the client
+  //   if there are no more retries available.
+  // * After per-try timeout, an error response would be discard, as a retry in the form of a hedged request is already in progress.
```
Comment (mpuncel) [inline api/envoy/api/v2/route/route_components.proto:1135]: s/discard/discarded

File: api/envoy/api/v2/route/route_components.proto [hunk 7]
```diff
@@ -1125,13 +1125,18 @@ message HedgePolicy {
...
+  //
+  // Note: For this to have effect, the :ref:`RetryPolicy <envoy_api_msg_route.RetryPolicy>` must be one that retries on timeout
+  // (e.g. `gateway-error`).
```
Comment (mpuncel) [inline api/envoy/api/v2/route/route_components.proto:1138]: i don't think this is true, you don't need to have `gateway-error` to have it be retried. A hedged per try timeout is always retried

Comment (ikonst) [reply api/envoy/api/v2/route/route_components.proto:1138]: @mpuncel 

I tested it and found it to be true. (In fact, wasted quite some time trying to understand why hedging doesn't work).

I actually wrote this above in reply to a similar question you raised:

Looking at this code, it seems that `retries_remaining_` would only get initialized to a non-zero value if `retry_on_` is set:
https://github.com/envoyproxy/envoy/blob/2709b6bfc73d089d823c5cde6c883806ec8f3d8e/source/common/router/retry_state_impl.cc#L121-L127

Thus `shouldRetry` would end up returning `RetryStatus::NoRetryLimitExceeded` for the hedged requests.

I can imagine this wasn't intended, since it surprised me too, but on the other hand it's congruent with how non-hedged retrying behaves.

Comment (mpuncel) [reply api/envoy/api/v2/route/route_components.proto:1138]: interesting! I do think that isn't intentional. In that case I'm not really sure how to word the comment, maybe you could say "you must have a RetryPolicy that retries at least one error code and specify the max number of retries". You don't have to have `gateway-error` specifically though it looks like from that code snippet

**Comment by snowp** [comment]
@ikonst I think this change looks good now. Could you a) merge master b) fix the formatting issue and c) apply the same change to the V3 docs (then run `tools/proto_format.sh fix` to propagate to v4alpha)? 

Also friendly reminder that force pushing breaks the reviewing flow for many, so avoid it if you can.

Thanks!

**Comment by ikonst** [comment]
I'm not a fan of rebases either since they break the already-reviewed / new-commits separation. (Perhaps I did this one because I forgot some sign-offs mid-way?)

**Comment by ikonst** [comment]
retest

**Comment by ikonst** [comment]
@snowp ^

**Review by snowp — APPROVED** [review]
LGTM, thanks!

@envoyproxy/api-shepherds for API review and V2 sign off
